File: src/cobalt/debug/backend/content/overlay_agent.js
// Copyright 2019 The Cobalt Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

(function(debugBackend) {

// Attach methods to handle commands in the 'Overlay' devtools domain.
// https://chromedevtools.github.io/devtools-protocol/tot/Overlay
let commands = debugBackend.Overlay = {};

// Returns non-overlapping rectangles to highlight for the box model of a node.
// This pseudo-command in the Overlay domain is a helper for the C++
// |OverlayAgent::HighlightNode|.
commands._highlightNodeRects = function(params) {
  let node = debugBackend.DOM._findNode(params);
  let config = params.highlightConfig || {};
  let highlights = [];
  if (node && node.getBoundingClientRect) {
    let styles = window.getComputedStyle(node);
    let content = node.getBoundingClientRect();
    let color;
    let box;

    let transformed = isNodeTransformed(node);

    if (!transformed) {
      // Margin
      color = config.marginColor;
      box = styleBox(styles, 'margin');
      if (color) {
        boxRects(content, box).forEach(
            rect => highlights.push(highlightParams(rect, color)));
      }

      // Border
      color = config.borderColor;
      box = styleBox(styles, 'border');
      if (color) {
        boxRects(content, box).forEach(
            rect => highlights.push(highlightParams(rect, color)));
      }
      content = insetRect(content, box);

      // Padding
      color = config.paddingColor;
      box = styleBox(styles, 'padding');
      if (color) {
        boxRects(content, box).forEach(
            rect => highlights.push(highlightParams(rect, color)));
      }
      content = insetRect(content, box);
    }

    // Content
    color = config.contentColor;
    if (color) {
      let highlight = highlightParams(content, color);
      if (transformed) {
        highlight.outlineColor = {r: 255, g: 0, b: 255, a: 1.0};
      }
      highlights.push(highlight);
    }
  }
  return JSON.stringify({highlightRects: highlights});
}

// Returns the inset width of the 4 sides of a box in the computed style, with
// margin being negative to place it outside the rect.
function styleBox(styles, boxName) {
  let suffix = boxName == 'border' ? '-width' : '';
  let sign = boxName == 'margin' ? -1 : 1;
  let box = {};
  ['top', 'right', 'bottom', 'left'].forEach(side => {
    let width = styles.getPropertyValue(`${boxName}-${side}${suffix}`);
    box[side] = sign * parseFloat(width) || 0;
  });
  return box;
}

function isNodeTransformed(node) {
  while (node) {
    if (window.getComputedStyle(node).transform !== 'none') return true;
    node = node.offsetParent;
  }
  return false;
}

// Returns an array of non-overlapping rectangles for the box around the inside
// of |rect|, but the rectangle for any box side with negative width will be
// on the outside of |rect|.
function boxRects(rect, box) {
  // Start out assuming box widths are all positive.
  let outerT = rect.y;
  let outerB = rect.y + rect.height;
  let outerL = rect.x;
  let outerR = rect.x + rect.width;
  let innerT = outerT + box.top;
  let innerB = outerB - box.bottom;
  let innerL = outerL + box.left;
  let innerR =  outerR - box.right;

  // Swap any inner/outer "inverted" by a negative box side.
  if (outerT > innerT) [outerT, innerT] = [innerT, outerT];
  if (outerB < innerB) [outerB, innerB] = [innerB, outerB];
  if (outerL > innerL) [outerL, innerL] = [innerL, outerL];
  if (outerR < innerR) [outerR, innerR] = [innerR, outerR];

  // +--------------+
  // |              |
  // +--+--------+--+
  // |  |        |  |
  // |  |        |  |
  // +--+--------+--+
  // |              |
  // +--------------+
  return [
    // top
    { x: outerL,
      y: outerT,
      width: outerR - outerL,
      height: innerT - outerT },
    // bottom
    { x: outerL,
      y: innerB,
      width: outerR - outerL,
      height: outerB - innerB },
    // left
    { x: outerL,
      y: innerT,
      width: innerL - outerL,
      height: innerB - innerT },
    // right
    { x: innerR,
      y: innerT,
      width: outerR - innerR,
      height: innerB - innerT },
  ];
}

// Returns the rectangle with the box around the insides removed.
function insetRect(rect, box) {
  return {
    x: rect.x + box.left,
    y: rect.y + box.top,
    width: rect.width - box.left - box.right,
    height: rect.height - box.top - box.bottom };
}

// Returns parameters matching the DevTools protocol "Overlay.highlightRect"
// parameters, as expected by the native overlay agent.
function highlightParams(rect, color) {
  // Copy each property rather than whole objects to ensure this can be
  // converted with JSON.serialize().
  return {
    x: rect.x,
    y: rect.y,
    width: rect.width,
    height: rect.height,
    color: {
      r: color.r,
      g: color.g,
      b: color.b,
      a: color.a,
    },
  };
}

// TODO: Pass debugBackend from C++ instead of getting it from the window.
})(window.debugBackend);

File: cobalt/debug/backend/overlay_agent.cc
// Copyright 2019 The Cobalt Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "cobalt/debug/backend/overlay_agent.h"

#include "cobalt/math/clamp.h"
#include "cobalt/math/rect_f.h"
#include "cobalt/render_tree/brush.h"
#include "cobalt/render_tree/color_rgba.h"
#include "cobalt/render_tree/composition_node.h"
#include "cobalt/render_tree/rect_node.h"

namespace cobalt {
namespace debug {
namespace backend {

using base::Value;
using math::Clamp;
using render_tree::ColorRGBA;

namespace {
// Returns the float value of a param, or 0.0 if undefined or non-numeric.
float GetFloatParam(const Value* params, base::StringPiece key) {
  if (!params || !params->is_dict()) return 0.0f;
  const Value* v = params->FindKey(key);
  if (!v || !(v->is_double() || v->is_int())) return 0.0f;
  return static_cast<float>(v->GetDouble());
}

// Returns an RGBA color defined by a param, or transparent if undefined.
ColorRGBA RenderColor(const Value* params) {
  float r = GetFloatParam(params, "r") / 255.0f;
  float g = GetFloatParam(params, "g") / 255.0f;
  float b = GetFloatParam(params, "b") / 255.0f;
  float a = GetFloatParam(params, "a");
  return ColorRGBA(Clamp(r, 0.0f, 1.0f), Clamp(g, 0.0f, 1.0f),
                   Clamp(b, 0.0f, 1.0f), Clamp(a, 0.0f, 1.0f));
}

// Returns a rectangle to render according to the params for the DevTools
// "Overlay.highlightRect" command.
// https://chromedevtools.github.io/devtools-protocol/tot/Overlay#method-highlightRect
scoped_refptr<render_tree::RectNode> RenderHighlightRect(const Value* params) {
  float x = GetFloatParam(params, "x");
  float y = GetFloatParam(params, "y");
  float width = GetFloatParam(params, "width");
  float height = GetFloatParam(params, "height");
  ColorRGBA color(RenderColor(params->FindKey("color")));
  const Value* outline_param = params->FindKey("outlineColor");
  ColorRGBA outline_color(RenderColor(outline_param));
  float outline_width = outline_param ? 1.0f : 0.0f;
  return base::MakeRefCounted<render_tree::RectNode>(
      math::RectF(x, y, width, height),
      std::make_unique<render_tree::SolidColorBrush>(color),
      std::make_unique<render_tree::Border>(render_tree::BorderSide(
          outline_width, render_tree::kBorderStyleSolid, outline_color)));
}

}  // namespace

OverlayAgent::OverlayAgent(DebugDispatcher* dispatcher,
                           std::unique_ptr<RenderLayer> render_layer)
    : AgentBase("Overlay", "overlay_agent.js", dispatcher),
      render_layer_(std::move(render_layer)) {
  DCHECK(render_layer_);

  commands_["highlightNode"] =
      base::Bind(&OverlayAgent::HighlightNode, base::Unretained(this));
  commands_["highlightRect"] =
      base::Bind(&OverlayAgent::HighlightRect, base::Unretained(this));
  commands_["hideHighlight"] =
      base::Bind(&OverlayAgent::HideHighlight, base::Unretained(this));
  commands_["setShowViewportSizeOnResize"] = base::Bind(
      &OverlayAgent::SetShowViewportSizeOnResize, base::Unretained(this));
}

void OverlayAgent::HighlightNode(Command command) {
  if (!EnsureEnabled(&command)) return;

  // Use the injected JavaScript helper to get the rectangles to highlight for
  // the specified node.
  JSONObject rects_response = dispatcher_->RunScriptCommand(
      "Overlay._highlightNodeRects", command.GetParams());
  const Value* highlight_rects =
      rects_response->FindPath({"result", "highlightRects"});
  if (!highlight_rects) {
    command.SendErrorResponse(Command::kInvalidParams,
                              "Can't get node highlights.");
    return;
  }

  // Render all the highlight rects as children of a CompositionNode.
  render_tree::CompositionNode::Builder builder;
  for (const Value& rect_params : highlight_rects->GetList()) {
    builder.AddChild(RenderHighlightRect(&rect_params));
  }
  render_layer_->SetFrontLayer(
      base::MakeRefCounted<render_tree::CompositionNode>(builder));

  command.SendResponse();
}

void OverlayAgent::HighlightRect(Command command) {
  if (!EnsureEnabled(&command)) return;

  JSONObject params = JSONParse(command.GetParams());
  render_layer_->SetFrontLayer(RenderHighlightRect(params.get()));
  command.SendResponse();
}

void OverlayAgent::HideHighlight(Command command) {
  if (!EnsureEnabled(&command)) return;

  render_layer_->SetFrontLayer(scoped_refptr<render_tree::Node>());
  command.SendResponse();
}

void OverlayAgent::SetShowViewportSizeOnResize(Command command) {
  // NO-OP implementation needed to prevent a frontend runtime exception when
  // Overlay settings are restored after the Performance panel stops tracing.
  command.SendResponse();
}

}  // namespace backend
}  // namespace debug
}  // namespace cobalt
